fdb-record-layer-geophile/src/main/java/com/apple/foundationdb/record/geophile/SpatialFunctionKeyExpression.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * SpatialFunctionKeyExpression.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.geophile;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.metadata.Key;
+import com.apple.foundationdb.record.metadata.expressions.FunctionKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecord;
+import com.geophile.z.Space;
+import com.geophile.z.SpatialObject;
+import com.geophile.z.spatialobject.d2.Point;
+import com.google.protobuf.Message;
+import org.locationtech.jts.io.ParseException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Spatial function key expressions.
+ */
+@API(API.Status.EXPERIMENTAL)
+public abstract class SpatialFunctionKeyExpression extends FunctionKeyExpression {
+
+    protected SpatialFunctionKeyExpression(@Nonnull String name, @Nonnull KeyExpression arguments) {
+        super(name, arguments);
+    }
+
+    @Nullable
+    protected abstract SpatialObject parseSpatialObject(@Nonnull Space space, @Nonnull Key.Evaluated arguments) throws ParseException;
+
+    @Nonnull
+    @Override
+    public <M extends Message> List<Key.Evaluated> evaluateFunction(@Nullable FDBRecord<M> record, @Nullable Message message, @Nonnull Key.Evaluated arguments) {
+        // TODO: How might we really get this out of the index maintainer? From the evaluation context?
+        //  Or should it be part of the key expression itself?
+        Space space = SpatialIndexMaintainer.SPACE_LAT_LON;
```

Comment:
I moved it into the key expression (function). I still didn't add a way to get a different space. Perhaps from the function name, so that there is a naming convention to distinguish geospatial from regular spatial?

Reply:
Yeah, I proposed in a different comment "geospatial" indexes on the globe and then perhaps "spatial" (or something else tbd later) for other spatial indexes as prefixes.

Reply:
It seems fine to me if it's one space per function